transaction/undoTx.go

```diff
@@ -154,40 +124,37 @@ func _initUndoTx(size int) *undoTx {
 }
 
 func NewUndoTx() TX {
-	if undoPool[0] == nil {
+	if undoPool == nil {
 		log.Fatal("Undo log not correctly initialized!")
 	}
-	t := <-undoPool[0]
-	return t
-}
-
-func NewLargeUndoTx() TX {
-	if undoPool[1] == nil {
-		log.Fatal("Undo log not correctly initialized!")
-	}
-	t := <-undoPool[1]
+	t := <-undoPool
 	return t
 }
 
 func releaseUndoTx(t *undoTx) {
 	t.abort()
-	if t.large {
-		undoPool[1] <- t
-	} else {
-		undoPool[0] <- t
-	}
+	undoPool <- t
 }
 
+// also takes care of increasing/decreasing the number of entries log can hold
 func (t *undoTx) updateLogTail(tail int) {
-	// atomic update
-	if t.large && tail >= LENTRYSIZE {
-		log.Fatal("Too large transaction. Already logged ",
-			LENTRYSIZE, " entries")
-	} else if !t.large && tail >= SENTRYSIZE {
-		log.Fatal("Too large transaction. Already logged ",
-			SENTRYSIZE, " entries")
+	if tail >= t.nEntry {
+		newE := 2 * t.nEntry // Double number of entries which can be stored
+		newLog := pmake([]entry, newE)
+		copy(newLog, t.log)
+		runtime.FlushRange(unsafe.Pointer(&newLog[0]),
+			uintptr(newE)*unsafe.Sizeof(newLog[0])) // Persist new log
+		t.log = newLog
```

Comment:
There has to be a store fence before updating `t.log`. Because we want update to `t.log` to be persisted only after all data updates made to `newLog` is fully persisted.